### DISCUSSION TITLE
How do I calculate MMFF energy with torsion constraints in 2022.03.2?

### DISCUSSION THREAD

**Discussion by k-khr** [rdkit/rdkit#5296] [Q&A] [answered]
I calculated the MMFF energy for different butane torsion angles as in https://github.com/rdkit/rdkit/issues/3781 in version 2022.03.2. The results are shown in the figure below, but It seems strange. Is there a problem with my code?

![butane 2022 03 2](https://user-images.githubusercontent.com/22492054/168573080-912406cc-77bd-422b-a18d-95c78ee244ee.png)

**Expected** (in 2019.03.3)
![butane 2019 03 3](https://user-images.githubusercontent.com/22492054/168573889-e3cd90e8-29eb-4219-a6d7-69b2c176a200.png)

There is code. I used butane.txt in https://github.com/rdkit/rdkit/issues/3781.

```py
from rdkit import Chem
from rdkit.Chem import AllChem
import numpy as np
import copy

m = Chem.rdmolfiles.MolFromMolFile('./butane.txt', removeHs = False)
mp = AllChem.MMFFGetMoleculeProperties(m)
globFF = AllChem.MMFFGetMoleculeForceField(m, mp)
m2 = copy.deepcopy(m)

angles = np.arange(-36., 36.) * 5.
energy = np.empty_like(angles)

for i, ang in enumerate(angles):
    mmff = AllChem.MMFFGetMoleculeForceField(m2, mp)
    mmff.MMFFAddTorsionConstraint(0, 1, 2, 3, False, ang - .1, ang + .1, 100000.0)
    mmff.Minimize()
    energy[i] = globFF.CalcEnergy( mmff.Positions() )
```

I installed rdkit by this command.

```sh
conda create -n rdkit_2022.03.2 -c conda-forge rdkit=2022.03.2 numpy matplotlib-base
```

**Comment by ptosco** [comment, marked as answer]
@k-khr There are two issues here.
The first is that you should set the dihedral to its target value ahead of the minimization, rather than trying to get the minimizer to do the job starting from a very different dihedral angle, as that may cause severe distortions of the geometry due to very large gradients.
The second is that you should not use a huge force constant, as that may ill-condition the minimizer and prevent it from converging on proper geometry.
If you use a force constant of 1000.0 the plot looks as one would expect:
```
from rdkit import Chem
from rdkit.Chem import AllChem
import numpy as np
import matplotlib.pyplot as plt

m = Chem.MolFromMolBlock("""Created with the Wolfram Language : www.wolfram.com
     WL             3D

 14 13  0  0  0  0  0  0  0  0999 V2000
   -1.5575    0.5684   -0.1320 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.6987   -0.6064    0.3102 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.6987   -0.6064   -0.3102 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.5575    0.5684    0.1320 C   0  0  0  0  0  0  0  0  0  0  0  0
   -1.6308    0.6129   -1.2232 H   0  0  0  0  0  0  0  0  0  0  0  0
   -2.5700    0.4662    0.2715 H   0  0  0  0  0  0  0  0  0  0  0  0
   -1.1524    1.5190    0.2272 H   0  0  0  0  0  0  0  0  0  0  0  0
   -1.2059   -1.5359    0.0253 H   0  0  0  0  0  0  0  0  0  0  0  0
   -0.6215   -0.6099    1.4037 H   0  0  0  0  0  0  0  0  0  0  0  0
    0.6215   -0.6099   -1.4037 H   0  0  0  0  0  0  0  0  0  0  0  0
    1.2059   -1.5359   -0.0253 H   0  0  0  0  0  0  0  0  0  0  0  0
    1.6308    0.6129    1.2232 H   0  0  0  0  0  0  0  0  0  0  0  0
    1.1524    1.5190   -0.2271 H   0  0  0  0  0  0  0  0  0  0  0  0
    2.5700    0.4662   -0.2715 H   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0
  2  3  1  0
  3  4  1  0
  1  5  1  0
  1  6  1  0
  1  7  1  0
  2  8  1  0
  2  9  1  0
  3 10  1  0
  3 11  1  0
  4 12  1  0
  4 13  1  0
  4 14  1  0
M  END
""", removeHs=False)
mp = AllChem.MMFFGetMoleculeProperties(m)
globFF = AllChem.MMFFGetMoleculeForceField(m, mp)

angles = np.arange(-36., 36.) * 5.
energy = np.empty_like(angles)

for i, ang in enumerate(angles):
    m2 = Chem.Mol(m)
    mmff = AllChem.MMFFGetMoleculeForceField(m2, mp)
    AllChem.SetDihedralDeg(m2.GetConformer(), 0, 1, 2, 3, ang)
    mmff.MMFFAddTorsionConstraint(0, 1, 2, 3, False, ang - .1, ang + .1, 1000.0)
    mmff.Minimize()
    energy[i] = globFF.CalcEnergy(mmff.Positions())

plt.plot(angles, energy)
```
![image](https://user-images.githubusercontent.com/5244385/168595210-f95f170f-8b6c-432f-9d4f-641e3cfa8a51.png)

You can see how the quality of the plot degrades when increasing the force constant to 10000.0:
![image](https://user-images.githubusercontent.com/5244385/168595648-c9a9ba4a-acf4-44ae-b698-d60df29411b4.png)

and 100000.0:
![image](https://user-images.githubusercontent.com/5244385/168595730-3fb3629d-091b-42c1-9a99-fcb7d3bce1e7.png)

**Reply by k-khr** [reply]
Thanks! It works in 2022.03.2.
